src/Jellyfin.MediaEncoding.Hls/Playlist/DynamicHlsPlaylistGenerator.cs

```diff
@@ -0,0 +1,204 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Jellyfin.MediaEncoding.Hls.Extractors;
+using Jellyfin.MediaEncoding.Keyframes;
+using MediaBrowser.Common.Configuration;
+using MediaBrowser.Controller.Configuration;
+using MediaBrowser.Controller.MediaEncoding;
+
+namespace Jellyfin.MediaEncoding.Hls.Playlist;
+
+/// <inheritdoc />
+public class DynamicHlsPlaylistGenerator : IDynamicHlsPlaylistGenerator
+{
+    private readonly IServerConfigurationManager _serverConfigurationManager;
+    private readonly IKeyframeExtractor[] _extractors;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DynamicHlsPlaylistGenerator"/> class.
+    /// </summary>
+    /// <param name="serverConfigurationManager">An instance of the see <see cref="IServerConfigurationManager"/> interface.</param>
+    /// <param name="extractors">An instance of <see cref="IEnumerable{IKeyframeExtractor}"/>.</param>
+    public DynamicHlsPlaylistGenerator(IServerConfigurationManager serverConfigurationManager, IEnumerable<IKeyframeExtractor> extractors)
+    {
+        _serverConfigurationManager = serverConfigurationManager;
+        _extractors = extractors.Where(e => e.IsMetadataBased).ToArray();
+    }
+
+    /// <inheritdoc />
+    public string CreateMainPlaylist(CreateMainPlaylistRequest request)
+    {
+        IReadOnlyList<double> segments;
+        if (TryExtractKeyframes(request.FilePath, out var keyframeData))
+        {
+            segments = ComputeSegments(keyframeData, request.DesiredSegmentLengthMs);
+        }
+        else
+        {
+            segments = ComputeEqualLengthSegments(request.DesiredSegmentLengthMs, request.TotalRuntimeTicks);
+        }
+
+        var segmentExtension = EncodingHelper.GetSegmentFileExtension(request.SegmentContainer);
+
+        // http://ffmpeg.org/ffmpeg-all.html#toc-hls-2
+        var isHlsInFmp4 = string.Equals(segmentExtension, "mp4", StringComparison.OrdinalIgnoreCase);
+        var hlsVersion = isHlsInFmp4 ? "7" : "3";
+
+        var builder = new StringBuilder(128);
```

Comment:
Is there a reason you give it an explicit size here? Given the foreach loop below it'll have to get bigger anyway, might as well leave the default up to the SDK right? (unless it's very stupid of course)

Reply:
Might be leftover from... something